### PR TITLE
Allow option set columns to be used with Sort and SortByColumns

### PR DESCRIPTION
**Pull request by sagarbakliwal**
Dataverse allows Option set (Choice) columns to be sorted. Only limitation is multiselect option sets (Choices) which are registered as Table and won't be allowed in Sort and SortByColumns

### CODE REVIEW

**Comment by lesaltzm** [comment]
Option sets have support in the interpreter, please add *.txt test cases covering the semantics of sorting on option set values.

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Sort.cs [hunk 1]
```diff
@@ -210,6 +210,9 @@ public static string GetSortComparatorIdForType(DType type)
 
             switch (type.Kind)
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Sort.cs:211]: Not necessarily as part of this change, but this should be pulled back to PA-Client, it only affects JSTranslator's impl of Sort/SortByColumns.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs [hunk 2]
```diff
@@ -441,10 +442,14 @@ public static async ValueTask<FormulaValue> SortTable(EvalVisitor runner, EvalVi
             {
                 return SortValueType<DateTimeValue, DateTime>(pairs, irContext, compareToResultModifier);
             }
-            else
+            else if (allDates)
             {
                 return SortValueType<DateValue, DateTime>(pairs, irContext, compareToResultModifier);
             }
+            else
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs:449]: else if (allOptionSets)?
And return an ErrorValue of some sort if no conditions match

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs [hunk 3]
```diff
@@ -476,6 +481,27 @@ private static bool IsValueTypeErrorOrBlank<T>(FormulaValue val)
             return new InMemoryTableValue(irContext, pairs.Select(pair => pair.row));
         }
 
+        private static FormulaValue SortOtionSet(List<(DValue<RecordValue> row, FormulaValue sortValue)> pairs, IRContext irContext, int compareToResultModifier)
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs:484]: sp: SortOptionSet

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Sort.txt [hunk 4]
```diff
@@ -1,4 +1,12 @@
-﻿>> Sort([-2, -1, 5, 1, 2], Value)
+﻿#SETUP: OptionSetSortTestSetup
+
+>> Sort(t1, OptionSetField1, SortOrder.Ascending)
+Table({OptionSetField1:OptionSet.option_1},{OptionSetField1:OptionSet.option_2})
```
Comment (lesaltzm) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Sort.txt:4]: could you add a test with multiple rows that have the same option set value?

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Sort.cs (hunk 1 above)
Comment (sagarbakliwal) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Sort.cs:211]: Included it as part of this change and added it to PA-Client pull request

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/PowerFxEvaluationTests.cs [hunk 5]
```diff
@@ -56,6 +57,46 @@ private static (RecalcEngine engine, RecordValue parameters) OptionSetTestSetup(
...
+            optionSet.TryGetValue(new DName("option_1"), out var o1Val);
+            optionSet.TryGetValue(new DName("option_2"), out var o2Val);
+
+            var r1 = FormulaValue.NewRecordFromFields(new NamedValue("OptionSetField1", o1Val), new NamedValue("StrField1", FormulaValue.New("test1")));
+            var r2 = FormulaValue.NewRecordFromFields(new NamedValue("OptionSetField1", o2Val), new NamedValue("StrField1", FormulaValue.New("test2")));
+            var r3 = FormulaValue.NewRecordFromFields(new NamedValue("OptionSetField1", o1Val), new NamedValue("StrField1", FormulaValue.New("test3")));
+            var r4 = FormulaValue.NewRecordFromFields(new NamedValue("OptionSetField1", o2Val), new NamedValue("StrField1", FormulaValue.New("test4")));
+            
+            // Testing with missing/blank option set field is throwing an exception
+            var r5 = FormulaValue.NewRecordFromFields(new NamedValue("StrField1", FormulaValue.New("test5")));
```
Comment (sagarbakliwal) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/PowerFxEvaluationTests.cs:79]: @MikeStall record r5 is missing option set field and it throws an exception Microsoft.PowerFx.Types.OptionSetValueType doesn't implement DefaultExpressionValue

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Sort.txt [hunk 6]
```diff
@@ -1,4 +1,24 @@
...
+>> Sort(t1, OptionSetField1, SortOrder.Descending)
+Table({OptionSetField1:OptionSet.option_2,StrField1:"test2"},{OptionSetField1:OptionSet.option_1,StrField1:"test1"})
+
+>> Sort(t2, OptionSetField1, SortOrder.Ascending)
+Table({OptionSetField1:OptionSet.option_1,StrField1:"test1"},{OptionSetField1:OptionSet.option_1,StrField1:"test3"},{OptionSetField1:OptionSet.option_2,StrField1:"test2"},{OptionSetField1:OptionSet.option_2,StrField1:"test4"})
+
+>> Sort(t2, OptionSetField1, SortOrder.Descending)
+Table({OptionSetField1:OptionSet.option_2,StrField1:"test2"},{OptionSetField1:OptionSet.option_2,StrField1:"test4"},{OptionSetField1:OptionSet.option_1,StrField1:"test1"},{OptionSetField1:OptionSet.option_1,StrField1:"test3"})
+
+>> Sort(t3, OptionSetField1, SortOrder.Ascending)
```
Comment (sagarbakliwal) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Sort.txt:15]: @MikeStall  This test fails due to blank option set issue

Comment (sagarbakliwal) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Sort.txt]: I have commented out the failing/unsupported test case for now

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/PowerFxEvaluationTests.cs (hunk 5 above)
Comment (sagarbakliwal) [reply src/tests/Microsoft.PowerFx.Interpreter.Tests/PowerFxEvaluationTests.cs:79]: Commented the failing/unsupported test case for now

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/PowerFxEvaluationTests.cs [hunk 7]
```diff
@@ -56,6 +57,46 @@ private static (RecalcEngine engine, RecordValue parameters) OptionSetTestSetup(
...
+            var r4 = FormulaValue.NewRecordFromFields(new NamedValue("OptionSetField1", o2Val), new NamedValue("StrField1", FormulaValue.New("test4")));
+            
+            // Testing with missing/blank option set field is throwing an exception. Once that is resolved uncomment and fix the test case in Sort.txt
+            var r5 = FormulaValue.NewRecordFromFields(new NamedValue("StrField1", FormulaValue.New("test5")));
+
+            var rType = RecordType.Empty()
+                .Add(new NamedFormulaType("OptionSetField1", FormulaType.OptionSetValue, "DisplayNameField1"))
+                .Add(new NamedFormulaType("StrField1", FormulaType.String, "DisplayNameField2"));
+
+            var t1 = new RecordsOnlyTableValue(Core.IR.IRContext.NotInSource(rType.ToTable()), new List<RecordValue>() { r1, r2 });
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/PowerFxEvaluationTests.cs:85]: >var t1 = new RecordsOnlyTableValue(Core.IR.IRContext.NotInSource(rType.ToTable()), new List<RecordValue>() { r1, r2 }); [](http://example.com/codeflow?start=12&length=119)

nit:  we have a public helper to simplify this
var t1 = FormulaValue.NewTable(rType, r1, r2);

(IRContext is internal)

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Sort.txt (hunk 6 above)
Comment (MikeStall) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Sort.txt]: This is hitting #731. 
But it's just happening in the test harness when trying to print the table  w/ option set. 
If we wrap in a Concat(), we avoid the test bug and still excercise the code. 

Can you adjust test case like this:
>> Concat(Sort(t1, OptionSetField1, SortOrder.Ascending), ThisRecord.StrField1 & ",")
"test1,test2,"

Then the output is a string, which the harness handles. 
I'd mention 731 in the test comments too.

File: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Sort.cs (hunk 1 above)
Comment (lesaltzm) [reply src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Sort.cs]: Thanks!

**Review by lesaltzm — DISMISSED** [review]
🚢

**Review by lesaltzm — APPROVED** [review]

**Pull request merged**